### ISSUE TITLE
todo: figure out how to instrument .mjs files

### ISSUE THREAD

**Issue by bcoe** [istanbuljs/nyc#659]
@bmeck has some amazing work close to landing that introduces ESM modules to node:

https://github.com/nodejs/node/pull/14369

Our strategy for hooking require statements (using require.extensions) will no longer work for `.mjs` files (which use the new ESM module syntax). @bmeck indicates that we might be able to pull off something similar using [import()](https://bugs.chromium.org/p/v8/issues/detail?id=6658&can=2&start=0&num=100&q=&colspec=ID%20Type%20Status%20Priority%20Owner%20Summary%20HW%20OS%20Component%20Stars&groupby=&sort=)

Let's try to stay in the loop, and make sure we have an implementation ready to go for `.mjs` as soon as it's possible.

**Comment by bmeck** [comment]
In newer builds of v8, code coverage is built in. I would say looking at https://chromedevtools.github.io/devtools-protocol/tot/Profiler/#method-startPreciseCoverage first might be the quickest approach.

Web compatible hooks for ESM are still being looked at and don't have a spec yet.

**Comment by bmeck** [comment]
~workflow:

1. `node --inspect-brk=$DEBUG_PORT app.js`
2. connect over `http://127.0.0.1:$DEBUG_PORT/json` to grab URL to send messages to over `ws:`
3. `Profiler.enable`
4. `Profiler.startPreciseCoverage`
5. `Runtime.enable` -> should generate a `Runtime.executionContextCreated` with id of `1`, that is the main context.
6. wait for `Runtime.executionContextDestroyed` with id of `1`.
7. `Profiler.takePreciseCoverage` to get dump of coverage

**Comment by jkrems** [comment]
Btw, there's a POC for implementing it here: https://github.com/bcoe/c8

**Comment by ORESoftware** [comment]
link to ESM?

**Comment by ORESoftware** [comment]
oh I thought it was WebAssembly for Node, but I guess it's ES6 modules for Node?

**Comment by OmgImAlexis** [comment]
Relating to what @bmeck was talking about https://github.com/nodejs/node/pull/22527 was just merged.

**Comment by OmgImAlexis** [comment]
Still an issue.

**Comment by OmgImAlexis** [comment]
Still an issue.

**Comment by jrgleason** [comment]
I am also looking to use NYC with ESM here is a[ SO](https://stackoverflow.com/questions/54682319/can-i-use-nyc-with-esm-and-mocha-without-the-std-esm-library). Is there a way to manually import the files for coverage?

**Comment by bcoe** [comment]
@coreyfarrell @jrgleason the ESM module implementation in Node.js completely overhauls how `require` works, and makes it so it's no longer possible to use some of nyc's tricks. we should be able to get `--experimental-modules` working in [c8](https://github.com/bcoe/c8) however, would love help doing so (adding tests would be a good start).

**Comment by coreyfarrell** [comment]
I just tested `node -r @babel/register --experimental-modules ./bin.mjs`, babel transformations did not happen.  I'm sure they'll eventually get that working but until they do I don't think we should make any effort to get it working with nyc.

**Comment by mikeal** [comment]
Is there a resolution to this yet?

@bcoe I tried using `c8` but it also doesn’t work yet :(

**Comment by Yaffle** [comment]
c8 works perfectly for me

**Comment by jrgleason** [comment]
@mikeal You are supposed to use mjs to signify it is a moduled javascript file. Another example is cjs for commonJs file. Eventually this may be enforced the same way extensions are required now so better to do it from the start ;-)

**Comment by coreyfarrell** [comment]
@jrgleason that's not true with `"type": "module"` in your package.json `.js` files are intepreted as ES modules.

That said the current status of node.js is that it doesn't support live transpilation of files loaded by the node.js native ES module loader (regardless of extension).  There are a couple proposals to address this but for now using nyc with native ES modules requires that you test pre-instrumented files.  As mentioned earlier the alternative is to use [c8](https://github.com/bcoe/c8) instead of nyc.

**Comment by vassudanagunta** [comment]
Is this necessary anymore given the recommended shift to ESM now that Node's support for it is no longer experimental?
